aioxmpp/avatar/service.py

```diff
@@ -0,0 +1,539 @@
+
+                sha1 = hashlib.sha1()
+                sha1.update(image_bytes)
+                id_computed = sha1.digest()
+                if id_ is not None:
+                    if id_ != id_computed:
+                        raise RuntimeError(
+                            "The given id does not match the SHA1 of "
+                            "the image data."
+                        )
+                else:
+                    id_ = id_computed
+
+                nbytes_computed = len(image_bytes)
+                if nbytes is not None:
+                    if nbytes != nbytes_computed:
+                        raise RuntimeError(
+                            "The given length does not match the length "
+                            "of the image data."
+                        )
+                else:
+                    nbytes = nbytes_computed
+
+                self._image_bytes = image_bytes
+                self._png_id = binascii.b2a_hex(id_).decode("ascii")
+
+        if nbytes is None:
+            raise RuntimeError(
+                "Image data length is not given an not inferable "
+                "from the other arguments."
+            )
+
+        if id_ is None:
+            raise RuntimeError(
+                "The SHA1 of the image data is not given an not inferable "
+                "from the other arguments."
+            )
+
+        if image_bytes is not None and mime_type != "image/png":
+            raise RuntimeError(
+                "The image bytes can only be given for image/png data."
+            )
+
+        if image_bytes is None and url is None:
+            raise RuntimeError(
+                "Either the image bytes or an url to retrive the avatar "
+                "image must be given."
+            )
+
+        self._metadata.info[mime_type].append(
+            avatar_xso.Info(
+                id_=id_, mime_type=mime_type, nbytes=nbytes,
+                width=width, height=height, url=url
+            )
+        )
+
+
+class AbstractAvatarDescriptor:
+    """
+    Description of an avatar source retrieved from pubsub.
+    """
+
+    def __init__(self, remote_jid, mime_type, id_, nbytes, width=None,
+                 height=None, url=None, pubsub=None):
+        self._remote_jid = remote_jid
+        self._mime_type = mime_type
+        self._id = id_
+        self._nbytes = nbytes
+        self._width = width
+        self._height = height
+        self._url = url
+        self._pubsub = pubsub
+
+    @asyncio.coroutine
+    def get_image_bytes(self):
+        """
+        Try to retrieve the image data corresponding to this avatar
+        descriptor.
+
+        This will raise :class:`NotImplementedError` if we are not
+        capable to retrieve the image data. It is guaranteed to not
+        raise :class:`NotImplementedError` if :attr:`image_in_pubsub`
+        is true.
+        """
+        raise NotImplementedError
+        yield
+
+    @property
+    def image_in_pubsub(self):
```

Comment:
Rename to ``is_image_in_pubsub`` or ``has_data_in_pubsub`` or something like that. It’s not clear that this is a boolean flag otherwise when reading code using it (I stumbled across this when reading the examples).